error/endpoints/error_collector.go

```diff
@@ -0,0 +1,49 @@
+package endpoints
+
+import "strings"
+
+// ErrorCollector provides collection of errors that occurred while evaluating
+// endpoint rule functions.
+type ErrorCollector struct {
```

Comment:
I'm not sure I'm understanding the need for this. Why would we not just return an error from any given rules function? Everytime we add an error to a collector in a rules function we bail and stop processing. This seems unnecessary and not idiomatic.

Reply:
That aside, I don't think `ErrorCollector` exists for any reason other than for us to check specific failure cases in testing. I'm not saying we should outright remove it, though.

Reply:
My understanding is that the purpose of the `ErrorCollector` is for more robust error logging and more robust test cases. For example, `isValidHostLabel`, the SEP states that this function is required to return a `bool`. While bailing on `false` would be acceptable (as far as the SEP is concerned), the caller would get no further context as to why this thing failed (at the level in which we call `isValidHostLabel`). the `ErrorCollector` provides the caller (either the actual consumer or a test case) with context as to why it failed.
Additionally, while I agree that passing a separate object for error reporting purposes may not be idiomatic, this is due to requirements imposed on us by the EP20 SEP. And, I think consumers may accept a hit on language-idiom compliance in order to get better error reporting.
And while, the `ErrorCollector` may not be required to write unit tests, without it, we wouldnt be able to ensure that a particular branch was tested (i.e. an input failed for a particular reason and not just failed for whatever reason).

Reply:
With that said, if necessary, this can totally be revisited as more of the implementation gets built out -- since this is an incremental merge into a feature branch.

Reply:
removed error collector now per standup discussion, will revisit later if i start to feel strongly about it